src/env/tock/storage.rs

```diff
@@ -58,85 +61,98 @@ mod memop_nr {
 }
 
 mod storage_type {
-    pub const STORE: usize = 1;
-    pub const PARTITION: usize = 2;
+    pub const STORE: u32 = 1;
+    pub const PARTITION: u32 = 2;
 }
 
-fn get_info(nr: usize, arg: usize) -> StorageResult<usize> {
-    let code = syscalls::command(DRIVER_NUMBER, command_nr::GET_INFO, nr, arg);
-    code.map_err(|_| StorageError::CustomError)
+fn get_info<S: Syscalls>(nr: u32, arg: u32) -> StorageResult<u32> {
```

Comment:
Should we put these `syscall` functions in a struct similar to those in `third_party/libtock-drivers/src/`?

```Rust
pub trait Config: platform::allow_ro::Config + platform::subscribe::Config
{
}

impl<T: platform::allow_ro::Config + platform::subscribe::Config> Config for T
{
}

pub struct Storage<S: Syscalls + RawSyscalls, C: Config = DefaultConfig>(S, C);

impl<S: Syscalls + RawSyscalls, C: Config> Storage<S, C> {
    pub fn get_info(nr: u32, arg: u32) -> StorageResult<u32> {
        let info = S::command(DRIVER_NUMBER, command_nr::GET_INFO, nr, arg)
            .to_result::<u32, ErrorCode>()
            .map_err(|_| StorageError::CustomError)?;
        Ok(info)
    }

    pub fn memop(nr: u32, arg: u32) -> StorageResult<u32> {
        let registers = unsafe { S::syscall2::<{ syscall_class::MEMOP }>([nr.into(), arg.into()]) };

        let r0 = registers[0].as_u32();
        let r1 = registers[1].as_u32();

        // make sure r0 is the `success with u32` (129) return variant and then return the value in r1 as u32
        // see: https://github.com/tock/tock/blob/master/doc/reference/trd104-syscalls.md#32-return-values
        match (r0, r1) {
            (129, r1) => Ok(r1),
            (_, _) => Err(StorageError::CustomError),
        }
    }

    pub fn write_slice(
        ptr: usize,
        value: &[u8],
    ) -> StorageResult<()> {
        share::scope(|allow_ro| {
            S::allow_ro::<C, DRIVER_NUMBER, { ro_allow_nr::WRITE_SLICE }>(allow_ro, value)
                .map_err(|_| StorageError::CustomError)?;
            block_command::<S, C>(
                DRIVER_NUMBER,
                command_nr::WRITE_SLICE,
                ptr as u32,
                value.len() as u32,
            )
        })
    }

    pub fn erase_page(
        ptr: usize,
        page_length: usize,
    ) -> StorageResult<()> {
        block_command::<S, C>(
            DRIVER_NUMBER,
            command_nr::ERASE_PAGE,
            ptr as u32,
            page_length as u32,
        )
    }

    fn block_command(
        driver: u32,
        cmd: u32,
        arg1: u32,
        arg2: u32,
    ) -> StorageResult<()> {
        let called: Cell<Option<(u32,)>> = Cell::new(None);

        share::scope(|subscribe| {
            S::subscribe::<_, _, C, DRIVER_NUMBER, { subscribe_nr::DONE }>(subscribe, &called)
                .map_err(|_| StorageError::CustomError)?;
            S::command(driver, cmd, arg1, arg2)
                .to_result::<(), ErrorCode>()
                .map_err(|_| StorageError::CustomError)?;
            libtock_drivers::util::Util::<S>::yieldk_for(|| called.get().is_some());
            if called.get().unwrap().0 == 0 {
                Ok(())
            } else {
                Err(StorageError::CustomError)
            }
        })
    }
}

unsafe fn read_slice(address: usize, length: usize) -> &'static [u8] {
    core::slice::from_raw_parts(address as *const u8, length)
}
```

Reply:
What's the benefit? The calls to these functions are not getting shorter, and the functions are private, so it's not really an API design advantage?

Reply:
I was picturing we would end up putting this in `third_party/libtock-drivers/src`. Is there a reason it is kept here?